### ISSUE TITLE
Relationship between .put(), .post() and .bulkDocs

### ISSUE THREAD

**Issue by jhs** [pouchdb/pouchdb#3916]
Hi! Are all update methods (.put(), .post(), .remove(), etc.) guaranteed to execute bulkDocs() ultimately, or is that an implementation detail, subject to change? If the latter, then may I suggest documentation or advice to plugin authors about how to intercept updates.

(I'm happy to submit that PR once I understand the answer.)

## Longer version

Currently, put(), post(), and other methods are implemented internally with calls to bulkDocs  (Only one document is in the batch.)

* post: https://github.com/pouchdb/pouchdb/blob/3.6.0/lib/adapter.js#L170
* put: https://github.com/pouchdb/pouchdb/blob/3.6.0/lib/adapter.js#L212
* remove: https://github.com/pouchdb/pouchdb/blob/3.6.0/lib/adapter.js#L311
* removeAttachment: https://github.com/pouchdb/pouchdb/blob/3.6.0/lib/adapter.js#L276
* replicator: https://github.com/pouchdb/pouchdb/blob/3.6.0/lib/replicate/replicate.js#L61

I wrote a plugin to intercept updates, for validation. I was shocked, shocked! to read #1412. No validate_doc_update? What? What! Have you all lost your minds? But then after some work and consideration, I got religion. That is a fantastic judgment call. Plugin-based validation is both simpler and more powerful. Bravo!

Unfortunately, now I need to be sure to validate all updates. So, that's my question. I works now with a single bulkDocs interception. However, if that changes in the future (say, a 3.x release), then I expose my database to invalid data.

I could intercept every possible method (put, post, removeAttachment), at a cost of much more complexity to my plugin (and potentially double-execution of the validator for those methods which call into bulkDocs).

In my opinion, PouchDB should eschew purity and acknowledge what is happening: bulkDocs is the workhorse of PouchDB. It will remain the workhorse for, say, the 3.x series. Look here! A custom handler to deal with this very fact:

https://github.com/pouchdb/pouchdb/blob/3.6.0/lib/adapter.js#L25-L34

Thus, documentation for plugin authors should say as much: "To intercept all updates, simply implement your own bulkDocs() method." This would also tie up the loose end in #1412, where people hope for some sort of pre-commit hook. bulkDocs() *is* that pre-commit hook.

Also, to bolster my argument, I would observe that the LevelDB community has had great success with this trick. For example, level-sublevel (a package used by PouchDB) does exactly this to simulate namespaces in a Level DB. level-sublevel just monkey-patches the .put() and .batch() methods. It works great.

Thank you very much, if you've read this far down!

**Comment by nolanlawson** [comment]
100% agree. I'd be happy to make our use of `bulkDocs()` official. It really simplifies just about everything in the Pouch codebase, and `pouchdb-validation` picks up the slack for `validate_doc_update` within the safety of plugin-land.

**Comment by marten-de-vries** [comment]
This is actually something of an open bug in ``pouchdb-validation``, currently a lot/all? checks are executed twice. Since those checks are stateless it's not that much of a problem (hence why it was never fixed), but it would be nice if I could reliably just wrap ``bulkDocs()``.

**Comment by jhs** [comment]
Hi, @marten-de-vries. Your code is what inspired me!

Great, @nolanlawson! I will submit a pull request reflecting this in the documentation. But I think, looking at the larger community, we may want a plugin too to give users very simple yes/no sync and async validation. The reason is because there is a bit of tedious logic required to do validation in bulkDocs(). One must validate all docs, and then pass the remaining ones to PouchDB for storage, but then re-assemble them so that the result array matches up with the `docs` parameter.

In other words:

    // User does this
    bulkDocs([{valid-1}, {invalid-2}, {valid-3}, {invalid-4}])

    // But actually PouchDB sees this
    bulkDocs([{valid-1}, {valid-2}])

    // But then the user gets this result
    [{ok:true}, {error:forbidden}, {ok:true}, {error:forbidden}]

**Comment by marten-de-vries** [comment]
@jhs I've got a couple of modules that try to simplify that (and ``pouchdb-validation`` uses them internally). [pouchdb-wrappers](https://www.npmjs.com/package/pouchdb-wrappers) & [pouchdb-bulkdocs-wrapper](https://www.npmjs.com/package/pouchdb-bulkdocs-wrapper). I think the last has a bug though concerning result ordering (although *technically* ordering isn't guaranteed anywhere anyway). Really need to document those plug-ins more, it's on my todo list.

**Comment by jhs** [comment]
Hmm. @nolanlawson Have you got any ideas or hints about where such documentation would go? I vaguely remember documentation about getting started writing a plugin but now I see very little.

Perhaps begin a fledgling section in doc/_includes/api/plugins.html? Some sort of "how to" listing, beginning with "how to intercept all writes"?

Thanks for any feedback!

**Comment by nolanlawson** [comment]
fixed by https://github.com/pouchdb/pouchdb/commit/1a047346ddf361600b094a1a5699a77e2eb1eef2
